### PR TITLE
Introduce marker files for download

### PR DESCRIPTION
**Pull request by charleso**
/cc @nhibberd @tmcgilchrist

### CODE REVIEW

File: src/Ambiata/Cli/Downloads.hs [hunk 1]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
+{-
+The layout of the DOWNLOAD_DIR looks like the following:
+
+  DOWNLOAD_DIR/
+    .marker/
+      file1
+      file2
+    file2
+
+Marker files are created when files are downloaded, to ensure they don't don't
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: don't don't

File: src/Ambiata/Cli/Downloads.hs [hunk 2]
```diff
@@ -58,10 +92,35 @@ doDownload dir a = do
...
+-- Remove any markers no longer contained in the list of server files.
+-- This assumes that the list of files is complete, and would not work
+--
+removeOldMarkers :: DownloadDir -> [ServerFile] -> IO [LocalFile]
+removeOldMarkers d a = do
+  let names = fmap serverFilePath a
+  let ignoreSystem = not . flip elem [".", ".."]
+  getDirectoryContents (markerDir d) >>= \x -> fmap concat . for (filter ignoreSystem x) $ \f -> do
+    if not $ elem (takeFileName f) names
+      then removeFile ((markerDir d) </> f) >> (pure . pure . LocalFile . T.pack) f
```
Comment (markhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Probably want to ignore ENOENT here, I think Mr Neale, had a combinator to do it somewhere around here.

**Comment by nhibberd** [comment]
Need some conceptual help here :white_medium_small_square:

File: src/Ambiata/Cli/Downloads.hs [hunk 3]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
-doDownload :: DownloadDir -> ServerFile -> AWS LocalFile
-doDownload dir a = do
+doDownload :: DownloadDir -> ServerFile -> AWS (Either DownloadError LocalFile)
+doDownload dir a = runEitherT $ do
   let fp = serverFilePath a
   liftIO $ putStrLn $ "Downloading " <> fp
   -- NOTE: This will download to a temporary file and then atomically rename once complete
-  download (unServerFile a) (unDownloadDir dir </> fp)
+  handle (\e -> case e of {
+      DestinationFileExists _ -> left $ DownloadFileExists (LocalFile $ T.pack fp)
```
Comment (markhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Don't think this is safe. Might need to discuss this case and what to do.

File: src/Ambiata/Cli/Downloads.hs [hunk 4]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
+    ; SourceMissing _ _ -> left $ ServerFileDoesNotExist a
+    ; e' -> throwM e'
+    })
+    . lift $ download (unServerFile a) (unDownloadDir dir </> fp)
```
Comment (markhibberd) [inline src/Ambiata/Cli/Downloads.hs]: @nhibberd Correct me if I am wrong, but I don't think you will be able to call this right? I don't think there are enough permissions.

File: src/Ambiata/Cli/Downloads.hs [hunk 5]
```diff
@@ -58,10 +92,35 @@ doDownload dir a = do
...
-  exists' <- doesFileExist $ d </> serverFilePath f
-  pure . not $ exists'
+isFileMissing d f =
+  fmap not . doesFileExist $ markerFile d f
+
+-- |
+-- Remove any markers no longer contained in the list of server files.
+-- This assumes that the list of files is complete, and would not work
+--
+removeOldMarkers :: DownloadDir -> [ServerFile] -> IO [LocalFile]
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Might be easier to remove all files older than 7 days? /thought

File: src/Ambiata/Cli/Downloads.hs [hunk 6]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
+    .marker/
+      file1
+      file2
+    file2
+
+Marker files are created when files are downloaded, to ensure they don't
+need to be downloaded again.
+
+This allows for users to manage the list of downloaded files separately without
+impacting the behaviour this program.
```
Comment (olorin) [inline src/Ambiata/Cli/Downloads.hs]: s/r t/r of t/

File: src/Ambiata/Cli/Downloads.hs [hunk 7]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
 -- Download any files from the remote dir that are not existing locally.
 --
-downloadFiles :: DownloadDir -> [ServerFile] -> AWS DownloadResult
+downloadFiles :: DownloadDir -> [ServerFile] -> AWS (Either DownloadError DownloadResult)
 downloadFiles dir a = do
+  liftIO $ createDirectoryIfMissing True (markerDir dir)
+  void . liftIO $ removeOldMarkers dir a
   toDo <- liftIO $ filesToDownload dir a
-  fmap DownloadResult $ mapM (doDownload dir) toDo
+  runEitherT . fmap DownloadResult $ mapM (EitherT . doDownload dir) toDo
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Not sure I see the point of running here only to wrap it back up on :62

File: src/Ambiata/Cli/Downloads.hs [hunk 8]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
+
 -- |
 -- Do the download with the given credentials
 --
 downloadReady :: DownloadDir -> Region -> DownloadAccess -> EitherT AmbiataError IO DownloadResult
 downloadReady dir r (DownloadAccess (TemporaryCreds k s sess) a) = do
   env <- setDebugging <$> getDebugging <*> newEnvFromCreds r k s (Just sess)
-  bimapEitherT AmbiataAWSDownloadError id . runAWS env $ downloadFiles dir a
+  runAWST env AmbiataAWSDownloadError
+    . bimapEitherT AmbiataDownloadError id . EitherT
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Would this be cleaner with just EitherT and a single error?

File: src/Ambiata/Cli/Downloads.hs [hunk 9]
```diff
@@ -29,28 +35,56 @@ import           System.Directory
...
 
-doDownload :: DownloadDir -> ServerFile -> AWS LocalFile
-doDownload dir a = do
+doDownload :: DownloadDir -> ServerFile -> AWS (Either DownloadError LocalFile)
+doDownload dir a = runEitherT $ do
   let fp = serverFilePath a
   liftIO $ putStrLn $ "Downloading " <> fp
   -- NOTE: This will download to a temporary file and then atomically rename once complete
-  download (unServerFile a) (unDownloadDir dir </> fp)
+  handle (\e -> case e of {
```
Comment (nhibberd) [inline src/Ambiata/Cli/Downloads.hs]: Exceptions are annoying

File: src/Ambiata/Cli/Downloads.hs [hunk 10]
```diff
@@ -58,10 +92,38 @@ doDownload dir a = do
...
 isFileMissing :: DownloadDir -> ServerFile -> IO Bool
-isFileMissing (DownloadDir d) f = do
-  exists' <- doesFileExist $ d </> serverFilePath f
-  pure . not $ exists'
+isFileMissing d f =
+  fmap not . doesFileExist $ markerFile d f
+
+-- |
+-- Remove any markers no longer contained in the list of server files.
+-- This assumes that the list of files is complete, and would not work
```
Comment (olorin) [inline src/Ambiata/Cli/Downloads.hs]: Was there supposed to be more to that last sentence?

**Comment by tmcgilchrist** [comment]
Looks pretty good to me :8ball:

File: src/Ambiata/Cli/Downloads.hs [hunk 11]
```diff
@@ -35,22 +59,32 @@ import           System.IO
...
+    $ downloadFiles dir a
 
 -- |
 -- Download any files from the remote dir that are not existing locally.
 --
-downloadFiles :: DownloadDir -> [ServerFile] -> AWS DownloadResult
+downloadFiles :: DownloadDir -> [ServerFile] -> AWS (Either DownloadError DownloadResult)
 downloadFiles dir a = do
+  liftIO $ createDirectoryIfMissing True (markerDir dir)
+  void . liftIO $ removeOldMarkers dir a
```
Comment (EduardSergeev) [inline src/Ambiata/Cli/Downloads.hs]: So we do remove markers files before downloading files. What happens if a user did move some files and then try to download the same directory again?
I am probably reading the intention of markers files incorrectly... but aren't they supposed to be kept for some time?

**Pull request merged**